src/cascadia/PublicTerminalCore/HwndTerminal.cpp

```diff
@@ -0,0 +1,413 @@
+
+    if (RegisterTermClass(hInstance))
+    {
+        _hwnd = wil::unique_hwnd(CreateWindowExW(
+            0,
+            term_window_class,
+            nullptr,
+            WS_CHILD |
+                WS_CLIPCHILDREN |
+                WS_CLIPSIBLINGS |
+                WS_VISIBLE,
+            0,
+            0,
+            0,
+            0,
+            parentHwnd,
+            nullptr,
+            hInstance,
+            nullptr));
+    }
+}
+
+HRESULT HwndTerminal::Initialize()
+{
+    _terminal = std::make_unique<::Microsoft::Terminal::Core::Terminal>();
+    auto renderThread = std::make_unique<::Microsoft::Console::Render::RenderThread>();
+    auto* const localPointerToThread = renderThread.get();
+    _renderer = std::make_unique<::Microsoft::Console::Render::Renderer>(_terminal.get(), nullptr, 0, std::move(renderThread));
+    RETURN_HR_IF_NULL(E_POINTER, localPointerToThread);
+    RETURN_IF_FAILED(localPointerToThread->Initialize(_renderer.get()));
+
+    auto dxEngine = std::make_unique<::Microsoft::Console::Render::DxEngine>();
+    RETURN_IF_FAILED(dxEngine->SetHwnd(_hwnd.get()));
+    RETURN_IF_FAILED(dxEngine->Enable());
+    _renderer->AddRenderEngine(dxEngine.get());
+
+    const auto pfn = std::bind(&::Microsoft::Console::Render::Renderer::IsGlyphWideByFont, _renderer.get(), std::placeholders::_1);
+    SetGlyphWidthFallback(pfn);
+
+    _UpdateFont(USER_DEFAULT_SCREEN_DPI);
+    RECT windowRect;
+    GetWindowRect(_hwnd.get(), &windowRect);
+
+    const COORD windowSize{ gsl::narrow<short>(windowRect.right - windowRect.left), gsl::narrow<short>(windowRect.bottom - windowRect.top) };
+
+    // Fist set up the dx engine with the window size in pixels.
+    // Then, using the font, get the number of characters that can fit.
+    // Resize our terminal connection to match that size, and initialize the terminal with that size.
+    const auto viewInPixels = Viewport::FromDimensions({ 0, 0 }, windowSize);
+    RETURN_IF_FAILED(dxEngine->SetWindowSize({ viewInPixels.Width(), viewInPixels.Height() }));
+
+    _renderEngine = std::move(dxEngine);
+
+    _terminal->SetBackgroundCallback([](auto) {});
+
+    _terminal->Create(COORD{ 80, 25 }, 1000, *_renderer);
+    _terminal->SetDefaultBackground(RGB(5, 27, 80));
+    _terminal->SetDefaultForeground(RGB(255, 255, 255));
+
+    localPointerToThread->EnablePainting();
+
+    return S_OK;
+}
+
+void HwndTerminal::RegisterScrollCallback(std::function<void(int, int, int)> callback)
+{
+    _terminal->SetScrollPositionChangedCallback(callback);
+}
+
+void HwndTerminal::RegisterWriteCallback(void _stdcall callback(wchar_t*))
+{
+    _terminal->SetWriteInputCallback([=](std::wstring & input) noexcept {
+        const wchar_t* text = input.c_str();
+        const size_t textChars = wcslen(text) + 1;
+        const size_t textBytes = textChars * sizeof(wchar_t);
+        wchar_t* callingText = nullptr;
+
+        callingText = static_cast<wchar_t*>(::CoTaskMemAlloc(textBytes));

```

Comment:
wil has a helper for this, actually. `wil::unique_cotaskmem_string`.

Also, we're never freeing this. Should we be? Does the callback absolutely take ownership of it?

Reply:
(I'd prefer you switch to the helper instead of adding a `CoTaskMemFree`, btw)

Reply:
I don't think we can use `wil::unique_cotaskmem_string` since that deallocates in its destructor. This function is designed to hand ownership of the string over to the caller, in this case that is .NET which will use `CoTaskMemFree` when the garbage collector decides that the string can be freed.

Reply:
That’s all the more reason to use the WIL one here: you can signal that ownership is handed over by using `.release` or `.detach` (I forget specifically which.)